google-ads-examples/src/main/java/com/google/ads/googleads/examples/hotelads/AddHotelListingGroupTree.java

```diff
@@ -0,0 +1,380 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.ads.googleads.examples.hotelads;
+
+import com.beust.jcommander.Parameter;
+import com.google.ads.googleads.examples.utils.ArgumentNames;
+import com.google.ads.googleads.examples.utils.CodeSampleParams;
+import com.google.ads.googleads.lib.GoogleAdsClient;
+import com.google.ads.googleads.v5.common.HotelClassInfo;
+import com.google.ads.googleads.v5.common.HotelCountryRegionInfo;
+import com.google.ads.googleads.v5.common.ListingDimensionInfo;
+import com.google.ads.googleads.v5.common.ListingGroupInfo;
+import com.google.ads.googleads.v5.enums.AdGroupCriterionStatusEnum.AdGroupCriterionStatus;
+import com.google.ads.googleads.v5.enums.ListingGroupTypeEnum.ListingGroupType;
+import com.google.ads.googleads.v5.errors.GoogleAdsError;
+import com.google.ads.googleads.v5.errors.GoogleAdsException;
+import com.google.ads.googleads.v5.resources.AdGroupCriterion;
+import com.google.ads.googleads.v5.services.AdGroupCriterionOperation;
+import com.google.ads.googleads.v5.services.AdGroupCriterionServiceClient;
+import com.google.ads.googleads.v5.services.MutateAdGroupCriteriaResponse;
+import com.google.ads.googleads.v5.services.MutateAdGroupCriterionResult;
+import com.google.ads.googleads.v5.utils.ResourceNames;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This example shows how to add a hotel listing group tree, which has two levels. The first level
+ * is partitioned by the hotel class. The second level is partitioned by the country region.
+ *
+ * <p>Each level is composed of two types of nodes: `UNIT` and `SUBDIVISION`. `UNIT` nodes serve as
+ * leaf nodes in a tree and can have bid amount set. `SUBDIVISION` nodes serve as internal nodes
+ * where a subtree will be built. The `SUBDIVISION` node cannot have bid amount set. See
+ * https://developers.google.com/google-ads/api/docs/hotel-ads/overview for more information.
+ *
+ * <p>Note: Only one listing group tree can be added per ad group. Attempting to add another listing
+ * group tree to an ad group that already has one will fail.
+ */
+public class AddHotelListingGroupTree {
+  /**
+   * The next temporary criterion ID to be used, which is a negative integer.
+   *
+   * <p>When creating a tree, we need to specify the parent-child relationships between nodes.
+   * However, until a criterion has been created on the server we do not have a criterion ID with
+   * which to refer to it.
+   *
+   * <p>Instead we can specify temporary IDs that are specific to a single mutate request. Once a
+   * criterion is created, it is assigned an ID as normal and the temporary ID will no longer refer
+   * to it.
+   */
+  private static int nextTempId = -1;
```

Comment:
After reading through the complete example, I think it might be more concise to use an `Iterator<Long>` like we do in the [AdWords API partition utility](https://github.com/googleads/googleads-java-lib/blob/master/modules/adwords_axis/src/main/java/com/google/api/ads/adwords/axis/utils/v201809/shopping/ProductPartitionTreeImpl.java#L69). This seems less error prone, since users don't have to remember to decrement a variable, and instead can just call `next()` on an iterator that supplies a decreasing sequence of negative longs.

WDYT?

Reply:
Agreed.